### DISCUSSION TITLE
find an image coordinates in another image

### DISCUSSION THREAD

**Discussion by gaful** [ImageMagick/ImageMagick#7353] [Help]
### ImageMagick version

7.1.1-33

### Operating system, version and so on

windows

### Description

Hello, 
I am trying to find an image in another image to then be able to crop the image for sending it tesseract. 
The parts of croping and sending to tesseract works fine. 
But I would like to have the crop the most precise possible to send only the relevant part to tesseract. 
My image are scanned, and the area to search is not at the exact same position at the same time. I have a subimage that is present in the source image and from its coordinates I would be able to locate the precise zone to search for. 
My issue is how to get the coordinates of my subimage. 
I have tryied the following 
`compare.exe -metric RMSE -subimage-search mainImage subImage NULL: 2>outputFile` 
But I have a lot of bad location : my image are black and white. Any idea on how to improve that ? 
I am probably taking it the wrong way. Any idea what to look for ? 
Many thanks 
G

**Comment by snibgo** [comment]
Please show sample input files.

**Reply by gaful** [reply]
Thank you @snibgo,
 Sure, here are example of mainImage an subImage : 
![mainImage](https://github.com/ImageMagick/ImageMagick/assets/33194277/511b91e7-f6f6-46d4-9dd3-f2c72f8d91fa)
![subImage](https://github.com/ImageMagick/ImageMagick/assets/33194277/f9aea8aa-cc77-4ffb-8ab1-7a77a7f0ddbe)
Once I have the coordinantes of the SubImage, I will be able to determine the coordinates of the area I want to OCR through tesseract.

**Comment by snibgo** [comment]
As you use IM v7, I suggest using the `magick` command. Naming your input files as `avisMain.png` and `avisSub.png`:
```
magick compare -metric RMSE -subimage-search avisMain.png avisSub.png NULL:
16329 (0.249165) @ 28,50
```

So the best match has the top-left corner of the subimage at (28,50) in the main image. The subimage is 117x120 pixels, so a crop could be:

```
magick avisMain.png -crop 117x120+28+50 +repage crp.png
```

The result, crp.png, looks reasonable:

![crp](https://github.com/ImageMagick/ImageMagick/assets/33812368/dd617a25-f3ea-44e6-ac14-dbdd0a88ec44)

**Reply by gaful** [reply]
Thank you very much @snibgo, it does decrease to almost 0 the false positive. 
But I do have still a lot of non found Subimage (~30%). Any advice or ressources to improve this ? 
The question I am looking at : 
* smaller or bigger subImage ?
* preprocessing on the image ?
 * any other idea ? 
Many thanks 
G

**Comment by snibgo** [comment]
> ... it does decrease to almost 0 the false positive. But I do have still a lot of non found Subimage (~30%).

I don't understand what that means. I have shown that I get only one match, and it is correct.

**Reply by gaful** [reply]
Sorry, what I meant is : 
I have thousands of image like the mainimage, some with little variation because they come from a scanner.: 
* sometimes the images are slightly tilted to the right or left  
* or little lighter or darker 
In those cases image magick does not found the subimage. A this time my matching ratio is around 60%, I would like to reach 80% or 90%. 
I was wondering if there was some best practices that you know of that I could look at to improve my matching ratio ? 
many thanks. 
G

**Reply by fmw42** [reply]
Template matching such as the magick compare command does not work when there is a difference in either scale, or rotation.  You might want to look at OpenCV feature matching.  There are scale and rotate invariant template matching techniques, but they are rather complex.

It might help us to see an example of your image and template that do not work.

**Comment by snibgo** [comment]
When text is scanned, minor rotations can be easily corrected, eg by trial-and-error, where each trial makes a small rotation, and measures the contrast when the result is scaled to a single column.

For more general but faster methods, see [What rotation?](http://im.snibgo.com/whatrot.htm) and [What rotation and scale?](http://im.snibgo.com/whatrotscale.htm) and [By FFT, what rotation?](http://im.snibgo.com/whatrotfft.htm).

For the "little lighter or darker" problem, I suggest normalising the inputs to the comparisons so they match in lightness, and perhaps also contrast. I think `-metric NCC` does this.

**Reply by fmw42** [reply]
If you have FFTW installed, then -metric NCC will be much faster than any of the other metrics.  There is also -deskew that might also help to unrotate the images.

**Comment by gaful** [comment]
many thanks, all discussion were very helpfull
